Diversity combining

MLsmd opened the pull request:
-Create C++ block diversity_combining_cc with SC capability. This closes #1.
-Add MRC capability. This closes #2 and closes #3.
-Add extensive QA tests. This closes #4.
-Create .xml file for GRC.
-Documentation for Doxygen.

MLsmd committed:
diversity_combining: Create c++ block and basic qa test for selection combining.

MLsmd committed:
diversity_combining: Expand qa test for SC.

MLsmd committed:
diversity_combining: Add MRC capability and QA test.

MLsmd committed:
diversity_combining: Change type of combining_technique from uint8_t to string for better readabiliy.

MLsmd committed:
diversity_combining: Add .xml file for GRC.

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+
+      // Collect all tags of the input buffer with key "csi" in the vector 'tags'.
+      std::vector <gr::tag_t> tags;
+      const std::string s = "csi";
```
Why are you doing this in the work function, if it never changes?

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+
+      uint16_t symbol_length; // Number of items in the current symbol.
+
+      // Handle the samples which are located before the first tag.
```
This comment should probably be located after this first if clause.

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+          }
+          // Get CSI from tag.
+          d_csi = pmt::c32vector_elements(tags[i].value);
+          // Calculated the weighting vector for the next symbol with the received CSI.
```
Typo in comment

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+      }
+      else if(d_combining_technique.compare("MRC") == 0) { // Maximum-Ratio combining
+        // Calculate the output stream as the weighted sum of the input streams.
+        std::fill(out, &out[length*d_vlen], 0.0);
```
Why are you using this notation? Just asking. I see why you might want to use std::fill instead of memset (even though I'm unsure about the performance), but you also use memcpy instead of std::copy.

MLsmd replied:
I think memset is not safe for floating point types, while memcpy can be used safely when you pass the proper length in bytes.
But I changed the memcpy now to std::copy to keep the coding style consistent.

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+    void
+    diversity_combiner_cc_impl::process_symbol(gr_vector_const_void_star input,
+                                               gr_complex* out,
+                                               uint16_t offset,
```
Why are offset and length just uint16_t? 64k is not small, but still in the range an application could get to. Better make this larger.

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+      // Handle the samples which are located before the first tag.
+      if(tags.size() == 0){ // Input buffer includes no tags at all.
+        symbol_length = noutput_items;
+        process_symbol(input_items, &out[nprocessed*d_vlen], nprocessed*d_vlen, symbol_length);
```
is nprocessed not always 0 at this point?

fwunsch commented on gr-digital/lib/diversity_combiner_cc_impl.cc:
```diff
@@ -0,0 +1,181 @@
+          /* There are items in the input buffer, before the first tag arrives,
+           * which belong to the previous symbol. */
+          symbol_length = tags[0].offset - nitems_read(0);
+          process_symbol(input_items, &out[nprocessed*d_vlen], nprocessed*d_vlen, symbol_length);
```
again: nprocessed == 0?

fwunsch reviewed (changes requested):
Looks really, really good so far! I like your coding style very much. I only added some small comments.

Two things crossed my mind: 
1) Wouldn't it be nice to make the combining method settable during runtime? This shouldn't be a major problem and would allow to instantly see the increase or decrease in performance connected to the respective scheme.

2) Have a look at std::valarray. If you are working a lot with vectors, this simplifies point-wise multiplication and other operations commonly required for DSP. This does not mean that I want you to use that everywhere, just have a look at it and decide or discuss if that's helpful for your work.

MLsmd committed:
diversity_combining: Minor improvements after code review.

MLsmd commented:
Thank you @fwunsch for your review!

- [x] The setter method is a great idea! I implemented it.
- [x] The std::valarray is an interesting class template for scientific vector calculation! After looking deeper into it, I decided not to use it. The std::valarray would simplify the code in 3 cases (reduce a for loop to one line of code). The initialization and transfer from the pmt vector to the std::valarray, however, would compensate this improvement. Additionially, the overload of the non-member function  std::begin(std::valarray) seems to be buggy and the code would be an ugly workaround.
- [x] I fixed the minor improvements in the code.

MLsmd committed:
diversity_combining: Change category of GRC block to 'MIMO' and add docu to GRC.
